### PR TITLE
Add markdown linting

### PR DESCRIPTION
**Pull request by dshevtsov**
## Purpose of this pull request

This pull request adds Markdown style check that is based on the https://github.com/markdownlint/markdownlint tool. The linter runs as a hook in serving mode for modified files only, and as a `rake lint` task for a specified path or for modified files if no path was provided.

To see the list of enabled rules, run:
```
bin/mdl --list-rules
```
or
```
bin/mdl --list-rules --show-aliases
```

The [rules](https://github.com/markdownlint/markdownlint/blob/master/docs/RULES.md) to be checked are set in the `_checks/md_style` file.

## Additional information

Internal ticket: MAGEDOC-3886

To test the implementation locally:

**Serving mode**:

1. Run `rake`
2. Change any `.md` file of the devdocs repository. The linter should report about broken rules unless content in the modified file mets the specified style.
3. Change another file. The linter should report about broken rules after regeneration is finished unless content in the modified file mets the specified style.
4. Stop jekyll (Ctrl-C)

**With `rake lint`**:

`rake lint` - The linter should report about broken rules in the modified files unless content in the modified files mets the specified style.

`rake lint path=mftf` - The linter should report about broken rules in files at the `mftf/` directory unless content of the files mets the specified style.

Example of the output:

```
~/Projects/repos/magento/devdocs (ds_lint-md *)$ rake lint
community/resources/best-practices.md:21: MD004 Unordered list style
community/resources/best-practices.md:22: MD004 Unordered list style
community/resources/best-practices.md:25: MD004 Unordered list style
community/resources/best-practices.md:26: MD004 Unordered list style
community/resources/best-practices.md:27: MD004 Unordered list style
community/resources/best-practices.md:28: MD004 Unordered list style
community/resources/best-practices.md:29: MD004 Unordered list style
community/resources/best-practices.md:30: MD004 Unordered list style
community/resources/best-practices.md:6: MD009 Trailing spaces
community/resources/best-practices.md:11: MD009 Trailing spaces
community/resources/best-practices.md:22: MD009 Trailing spaces
community/resources/best-practices.md:23: MD009 Trailing spaces
community/resources/best-practices.md:34: MD009 Trailing spaces
community/resources/best-practices.md:36: MD009 Trailing spaces
community/resources/best-practices.md:16: MD012 Multiple consecutive blank lines
community/resources/best-practices.md:32: MD026 Trailing punctuation in header

A detailed description of the rules is available at https://github.com/markdownlint/markdownlint/blob/master/docs/RULES.md
```

### CODE REVIEW

**Comment by hguthrie** [comment]
Will this check that my UL must be an asterisk?

**Comment by hguthrie** [comment]
Nevermind my question. I ran a test and it is honoring the UL style we use in the Cloud guide. so far, so good!

**Comment by keharper** [comment]
I object to the "no emphasis as headings" check. Practically every REST and GraphQL contains **Request** and **Response** pseudo-headings.

Also, the link to the rules in the description doesn't resolve. I need to know all the rules before I give my approval.

**Comment by hguthrie** [comment]
The rules are here: https://github.com/markdownlint/markdownlint/blob/master/docs/RULES.md

**Comment by dshevtsov** [comment]
@keharper I updated the link.

**Comment by dobooth** [comment]
This is going to require a good amount of initial cleanup. Plenty of "Headings ending with punctuation" errors out there. Likely a lot of "Expected asterisk, got dash" errors.
And we should suggest that writers are editing with the markdownlint extension installed so errors can be found before the rake stage. I know some are not. 
We can work on finessing the rules as needed.

**Comment by jeff-matthews** [comment]
This is a great first step in enforcing better markdown standards in our project. There's also a Visual Studio Code extension that enforces the same rules, which is helpful since several writers are already using the extension. It's not a requirement, but it's a bonus.

AS the team gets more acclimated to linting, we need to clean up our files and start enforcing this on Jenkins—much like we did with the link checker—and slowly increase coverage as we cleanup files.

**Comment by dshevtsov** [comment]
@keharper, to pass the [MD036](https://github.com/markdownlint/markdownlint/blob/master/docs/RULES.md#md036---emphasis-used-instead-of-a-header) check

```
**Response**

...
**Request**

...
```
should be changed to 
```
**Response:**

...

**Request:**

...

```
if you don't want to change them to headings.

I think this rule makes sense.

**Comment by dshevtsov** [comment]
running tests

**Comment by keharper** [comment]
I have many REST topics that contain multiple instances of the same heading. 

`guides/2.3/graphql/queries.md` contains a table that would not render correctly in markdown, so MD-033 fails

Somewhere, I've seen a URL flagged as failing a linter check, but it was inside a code sample and valid. Don't know where, though.

**Comment by dobooth** [comment]
Running the linter on the whole site returns this set of errors only (using the exclusion file in this PR):
01 Header levels should only increment by one level at a time
04 Unordered list style
05 Inconsistent indentation for list items at the same level
06 Consider starting bulleted lists at the beginning of the line
07 Unordered list indentation
09 Trailing spaces
10 Hard tabs
12 Multiple consecutive blank lines
22 Headers should be surrounded by blank lines
23 Headers must start at the beginning of the line
24 Multiple headers with the same content
26 Trailing punctuation in header
29 Ordered list item prefix
30 Spaces after list markers
31 Fenced code blocks should be surrounded by blank lines
32 Lists should be surrounded by blank lines
33 Inline HTML
36 Emphasis used instead of a header
40 Fenced code blocks should have a language specified
46 Code block style

File: Rakefile [hunk 1]
```diff
@@ -60,3 +60,17 @@ end
 
 desc 'Pull docs from external repositories'
 task init: %w[multirepo:init]
+
+desc 'Check Markdown syntax in modified files or in a particular file or directory by path (e.g. path=mftf)'
```
Comment (hguthrie) [inline Rakefile]: For some reason, I don't see this whole line when I access the help.

**Comment by hguthrie** [comment]
I ran this test on a couple of Cloud docker files with typical formatting. I LOVE that it picked up some problems that need to be addressed; however, the following items I had a problem with:

-  `Header levels should only increment by one level at a time`—I consider the Page Title to be the H1 of the page. This particular violation happened with a Cloud procedure title. All Cloud procedure titles are H4. We need a procedure title that does not show in the navigation and we need to be able to place it anywhere. Besides, if we just used bold text, it would ding us for Emphasis used instead of header, right?
-  `Trailing punctuation in header`—Caused by colon in the Procedure title format, which is '#### To do add a blah blah:'
-  `Spaces after list markers`—My linter expects at least two spaces after the list indicator. This makes it easier to read and lines up with numbered lists, too. We also work in YAML a lot, so my IDE is set to indent 4 spaces (not hard tab).
-  `Code block style`—I'm not sure what this was saying. It was found on a bash code block with appropriate fencing, indented 4 spaces inside a numbered list.
-  `Unordered list indentation`—Again, my IDE is set to indent 4 spaces (not hard tab). It is difficult to see indention of two spaces. 4 is more convenient since we frequently use YAML.
-  `Inline HTML`—It may be just a fluke, but this would sometimes show on line 1 of a file. ??

**Comment by dshevtsov** [comment]
YAML does not require 4 spaces. See [6.1. Indentation Spaces](https://yaml.org/spec/1.2/spec.html#id2777534).
> In YAML block styles, structure is determined by indentation. In general, indentation is defined as a zero or more space characters at the start of a line.

My lDE sets tabs as 2 spaces to follow Ruby conventions because I work a lot with Ruby. Magento uses a convention to use 4 spaces in PHP code, so many contributors may have 4 spaces or special tabs in their IDEs. So, different IDEs may have different settings.

Also, Markdown is space sensitive, because incorrect use of 4 and more spaces can turn the content to a code block unexpectedly.

Therefore, having a habit to use a simple space instead of a tab when you work on a Markdown file is a good compromise. It doesn't require to reconfigure the IDE, and allows to follow the existing Markdown convention.

**Comment by hguthrie** [comment]
I'm aware of the YAML requirement, but I am speaking about working with the Cloud YAML files. In any case, I do not use the indentations for code block. I use them in lists.  Markdown doesn't treat it like a code block when you use a dash first. There are many linters that expect two spaces, also.

I am not using tab. I uses spaces, as does my IDE.

**Comment by dshevtsov** [comment]
@hguthrie I excluded the rule because it has to be revisited by Docs team again in our next discussion.

**Comment by jeff-matthews** [comment]
Does anyone have any objections to merging this? Keep in mind, the list of rules is not final and will not affect Jenkins builds. Does anyone want more time to review the basic rake functionality?

**Comment by dshevtsov** [comment]
running tests

**Pull request merged**